### PR TITLE
Implementations of concepts and transformations discussed in #83

### PR DESCRIPTION
**Pull request by jrbeaumont**

### CODE REVIEW

File: src/Tuura/Concept/Circuit/Derived.hs [hunk 1]
```diff
@@ -171,6 +183,19 @@ xorGate :: a -> a -> a -> CircuitConcept a
 xorGate i1 i2 o = [rise i1, rise i2] ~|~> rise o <> [fall i1, fall i2] ~|~> rise o
                <> [rise i1, fall i2] ~|~> fall o <> [fall i1, rise i2] ~|~> fall o
 
+-- Set/reset latch including one non-inverted and one inverted output.
+srLatch :: Eq a => a -> a -> a -> a -> CircuitConcept a
+srLatch s r q nq = never [rise s, rise r]      -- disallow contradictory requests
+    <> initialise0 [q]   <> initialise1 [nq]   -- disallow initial state q=nq=0
```
Comment (snowleopard) [inline src/Tuura/Concept/Circuit/Derived.hs]: I think a better comment here would be `set initial state to q=0 and nq=1`. And add the following on top of the function:

`TODO: Generalise by allowing both 01 and 10 as initial states.`

File: src/Tuura/Concept/Circuit/Derived.hs [hunk 2]
```diff
@@ -89,41 +89,53 @@ dual c = mempty
...
 toggleSpecific a t
-    | signal t == a = toggle t
-    | otherwise     = t
+    | signal t `elem` a = toggle t
+    | otherwise         = t
 
 -- For the selected signal, toggle the directions of all transitions for this.
-bubbleCausality :: Eq a => a -> Causality (Transition a)
+bubbleCausality :: Eq a => [a] -> Causality (Transition a)
                         -> Causality (Transition a)
```
Comment (snowleopard) [inline src/Tuura/Concept/Circuit/Derived.hs]: Don't break this type signature, it's OK to occasionally exceed the line width limit.

File: src/Tuura/Concept/Circuit/Derived.hs [hunk 3]
```diff
@@ -89,41 +89,53 @@ dual c = mempty
...
 bubbleCausality s (Causality f t)
-    | signal t == s = Causality (map (toggleSpecific s) f) (toggle t)
-    | otherwise     = Causality (map (toggleSpecific s) f) t
+    | signal t `elem` s = Causality (map (toggleSpecific s) f) (toggle t)
+    | otherwise         = Causality (map (toggleSpecific s) f) t
 
 -- Invert the invariant transition of the selected signal.
-bubbleInvariant :: Eq a => a -> Invariant (Transition a)
+bubbleInvariant :: Eq a => [a] -> Invariant (Transition a)
                         -> Invariant (Transition a)
```
Comment (snowleopard) [inline src/Tuura/Concept/Circuit/Derived.hs]: Same here, let's not break this line.

**Comment by snowleopard** [comment]
@jrbeaumont I think the current version of `enable` is too strict: not only it restricts the behaviour of outputs, it can also prevent inputs from firing, which is probably undesirable. 

Perhaps, it would be a more useful function if it can be used in a more targeted way, for example by applying it only to specified output (or a set of outputs): `enable e [q, nq] $ srLatch s r q nq`. Imagine applying it to all signals in this case -- this would mean inputs can never fire!

It is similar to `bubble/bubbles` in the sense that it is commutative. But it is idempotent (enabling twice is the same as enabling once), whereas `bubble` is an involution, i.e. it's a self-inverse (inverting an input twice is a no-op).

**Comment by jrbeaumont** [comment]
@snowleopard: I didn't think of this! Thank you for pointing it out. Targeting it to specific signals would be a much better idea, or if possible, aiming it only at output/internal signals. However, this is much more difficult to do, as we cannot guarantee that the interface concepts will be included in the concept which is passed into `enable`.

p.s. idempotent and involution are fantastic words :)

**Comment by jrbeaumont** [comment]
@snowleopard I have made some changes, let me know what you think.

File: src/Tuura/Concept/Circuit/Derived.hs [hunk 4]
```diff
@@ -89,41 +89,52 @@ dual c = mempty
...
+bubbles :: Eq a => [a] -> CircuitConcept a -> CircuitConcept a
+bubbles s c = mempty
+    {
+       initial = bubbleInitialValue s (initial c),
+       arcs = fmap (bubbleCausality s) (arcs c),
+       interface = interface c,
+       invariant = fmap (bubbleInvariant s) (invariant c)
+    }
 
+-- Apply enable signal only for transitions of given signals in the concept
```
Comment (snowleopard) [inline src/Tuura/Concept/Circuit/Derived.hs]: I think in practice we almost always will apply `enable` only to a single output, so I'd prefer `enable` to take a single signal only. You can add `enables`, which sounds a bit odd, but is probably OK from the symmetry point of view :)

**Comment by snowleopard** [comment]
Wait. The new `enable t s c` can probably be implemented simply as `c <> (t ~> rise s <> t ~> fall s)`. What do you think?

**Comment by snowleopard** [comment]
Which brings me to another useful variant of `enable`, which we can call `sync`. It can be implemented similarly: `sync x y c = c <> (rise x ~> rise y <> fall x ~> fall y)`.

Simplifying even further: `sync x y c = c <> buffer x y`.

Now you can express `meElement` as `sync r1 g1 $ sync r2 g2 $ mutex g1 g2` :)

It doesn't look like a major step forward, but I think it's a fairly flexible higher-order transformation that can make some specifications easier to understand.

**Comment by snowleopard** [comment]
I realised there is a subtle difference between the current implementation and the one I proposed above.

Currently, `enable t s mempty == mempty`.

In the above proposal, `enable t s mempty = t ~> rise s <> t ~> fall s`.

I'm unsure which one is preferable. 

@jrbeaumont Any thoughts?

**Comment by jrbeaumont** [comment]
An enable for a single output can be done, it makes more sense to have it this way, but I would like to have `enables` like you said, for symmetry. 

>Wait. The new enable t s c can probably be implemented simply as c <> (t ~> rise s <> t ~> fall s).

This makes a lot more sense. An output will have constraints on both rise and fall transitions, so it only makes sense to generalise it, and make it more efficient, and ensure that the `enable` signal applies to both of it's transitions.

Sync seems interesting, I am not sure I quite understand it's use in this example. It seems to just replace buffer, but not use standard `<>` operator. Is it an idea to synchronise signals?

As far as for `mempty`, I am also unsure. I think ideally, if someone is to use `enable` on it's own, it should produce a result, it could be seen as a buffer in some cases, i.e. `s` indicates the polarity of `t`, so it is identifiable whether the circuit is enabled.

**Comment by jrbeaumont** [comment]
I have changed `enable`, and added `enables` and `sync`

**Comment by jrbeaumont** [comment]
The clearer `srLatch` implementation has been added.

**Comment by snowleopard** [comment]
@jrbeaumont Thanks, looks good!

**Pull request merged**